jina/optimizers/__init__.py

```diff
@@ -0,0 +1,100 @@
+from pathlib import Path
+from ruamel.yaml import YAML
+
+from .new_flow_runner import MultiFlowRunner
+from jina.helper import colored
+from jina.logging import default_logger as logger
+from .parameters import load_optimization_parameters
+
+
+class EvaluationCallback:
+    def __init__(self, eval_name=None):
+        self.op_name = eval_name
+        self.evaluation_values = {}
+        self.n_docs = 0
+
+    def flush(self):
+        self.evaluation_values = {}
+        self.n_docs = 0
+
+    def get_mean_evaluation(self):
+        if self.op_name:
+            return self.evaluation_values[self.op_name] / self.n_docs
+        return {
+            metric: val / self.n_docs for metric, val in self.evaluation_values.items()
+        }
+
+    def __call__(self, response):
+        self.n_docs += len(response.search.docs)
+        logger.info(f"Num of docs evaluated: {self.n_docs}")
+        for doc in response.search.docs:
+            for evaluation in doc.evaluations:
+                self.evaluation_values[evaluation.op_name] = (
+                    self.evaluation_values.get(evaluation.op_name, 0.0)
+                    + evaluation.value
+                )
+
+
+class OptunaOptimizer:
+    def __init__(
+        self,
+        index_flow_runner,
+        eval_flow_runner,
+        parameter_yaml,
+        best_config_filepath="config/best_config.yml",
+        workspace_env="JINA_WORKSPACE",
+    ):
+        self.index_flow_runner = index_flow_runner
+        self.eval_flow_runner = eval_flow_runner
+        self.parameter_yaml = parameter_yaml
+        self.best_config_filepath = Path(best_config_filepath)
+        self.workspace_env = workspace_env.lstrip("$")
+
+    def _trial_parameter_sampler(self, trial):
+        """https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.Trial.html#optuna.trial.Trial"""
+        trial_parameters = {}
+        parameters = load_optimization_parameters(self.parameter_yaml)
+        for param in parameters:
+            trial_parameters[param.env_var] = getattr(trial, param.method)(
+                **param.to_optuna_args()
+            )
+
+        trial_workspace = Path(
+            "JINA_WORKSPACE_" + "_".join([str(v) for v in trial_parameters.values()])
+        )
+        trial_parameters[self.workspace_env] = str(trial_workspace)
+
+        trial.workspace = trial_workspace
+        return trial_parameters
+
+    def _objective(self, trial):
+        self.eval_flow_runner.callback.flush()
```

Comment:
can't just a new callback be created? That would be more `functional-like`

Reply:
Passing the callback without init in the optimizer to recreate every time will require to pass the eval_name to Optimizer which I didn't like.

Reply:
I think is best practica not to reset the object

Reply:
but its okey

Reply:
I disagree. Resetting objects like this is a root for ugly and hard-to-track bugs. Please don't do that. There are two possibilities to solve that in general: 
1. Just give the eval_name to the `OptunaOptimizer`
2. The evaluator must implement a `empty_copy` function, which will return an empty copy of the evaluator. Guarantee, that custom evaluators implement it by defining an interface (named `EvaluationCallback`?), that evaluators must implement and inherit from this in the `MeanEvaluationCallback`

The first option is bad, as @bhavsarpratik already discovered, since it will not work with custom evaluator. Anyhow, the second is totally valid.

Reply:
I agree this shoukd be avoided but do not understand the evaluators copy. optimizer does not see the evaluator, just Flows containing Evaluators.